[WIP] TCPStore listen thread to execute callback function

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54265 [WIP] TCPStore listen thread to execute callback function**
* #54264 Add watch_key query and listen thread outline

WIP: CI failing in windows, need to use windows alternative to `::poll`

H-Huang committed:
TCPStore listen thread to execute callback function

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore listen thread to execute callback function"




[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -204,6 +273,15 @@ void TCPStoreDaemon::compareSetHandler(int socket) {
   } else {
     if (pos->second == currentValue) {
       pos->second = std::move(newValue);
+
```
Don't we need to handle `ADD` and `DELETE` as well?

H-Huang replied:
Good point, updated the PR

pritamdamania87 commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -39,15 +39,69 @@ enum class WatchResponseType : uint8_t { KEY_UPDATED };
+  tcputil::addPollfd(fds, storeListenSocket_, POLLIN);
+
+  while (1) {
+    SYSCHECK_ERR_RETURN_NEG1(::poll(fds.data(), fds.size(), -1));
```
On the server side, doesn't this poll interfere with the poll in TCPStoreDaemon?

H-Huang replied:
We now have to create a separate thread on the client to handle responses asynchronously from the master, this is running in that separate thread. I updated the first PR with more details.

pritamdamania87 reviewed (commented):
We probably need some unit tests for these changes. Also, its probably better to have changes like these in one PR instead of having part of them in https://github.com/pytorch/pytorch/pull/54264.

H-Huang committed:
Update on "[WIP] TCPStore listen thread to execute callback function"


WIP: CI failing in windows, need to use windows alternative to `::poll`

[ghstack-poisoned]

H-Huang committed:
Update on "[WIP] TCPStore listen thread to execute callback function"


WIP: CI failing in windows, need to use windows alternative to `::poll`

[ghstack-poisoned]

H-Huang committed:
Update on "[WIP] TCPStore listen thread to execute callback function"


WIP: CI failing in windows, need to use windows alternative to `::poll`

[ghstack-poisoned]

H-Huang committed:
Update on "[WIP] TCPStore listen thread to execute callback function"


WIP: CI failing in windows, need to use windows alternative to `::poll`

[ghstack-poisoned]

H-Huang committed:
Update on "[WIP] TCPStore listen thread to execute callback function"


WIP: CI failing in windows, need to use windows alternative to `::poll`

[ghstack-poisoned]

H-Huang commented:
Thanks for reviewing @pritamdamania87, closing this PR and combining the updated changes with #54264 as you mentioned
